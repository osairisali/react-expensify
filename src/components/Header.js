import React from "react";
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";
import { connect } from "react-redux";

export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify</h1>
    {/* perhatikan di sini juga perlu pake exact:{true} jika hanya redirect ke "/" */}
    <NavLink to="/dashboard" activeClassName="is-active">
      Dashboard
    </NavLink>
    <NavLink to="/create" activeClassName="is-active">
      Create Page
    </NavLink>
    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
