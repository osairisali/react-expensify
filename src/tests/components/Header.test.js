import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";
import Header from "../../components/Header";
import { shallow } from "enzyme";

// test suite pakai react-test-renderer package
test("should render Header correctly", () => {
  // renderer berfungsi untuk render react component yg bs dilihat
  const renderer = new ReactShallowRenderer();
  renderer.render(<Header />);

  // expect().toMatchSnapshot() untuk pertama kali akan membuat file snapshot hasil render
  // jika ada perubahan pada komponen yg berbeda dgn snapshot yg telah diambil, maka test fail
  // pada terminal bisa klik u untuk update snapshot sesuai perubahan
  expect(renderer.getRenderOutput()).toMatchSnapshot();
  // console.log(renderer.getRenderOutput())
});

// test suite yang sama dgn di atas, tp pakai Enzyme yg lbh user friendly & komprehensif
test("should render Header correctly", () => {
  // shallow berfungsi untuk render react component tidak sampai pada komponen children
  const wrapper = shallow(<Header />);

  // pada jest.config.json ada konfigurasi "enzyme-to-json/serializer" yg menyederhanakan
  // hasil snapshot pada Header.test.js.snap agar cm menampilkan bagian komponen react tanpa internal enzyme

  // raf sendiri adalah request animation frame untuk mensimulasikan animasi browser pada test suite

  expect(wrapper.find("h1").text()).toBe("Expensify");
});
