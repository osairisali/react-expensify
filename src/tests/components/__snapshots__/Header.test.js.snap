// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="button button__logout"
        onClick={[MockFunction]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
